### DISCUSSION TITLE
DevilutionX package in the Debian (+ Ubuntu, etc.) official repositories

### DISCUSSION THREAD

**Discussion by vv221** [diasurgical/DevilutionX#7556] [General]
I plan to start working soon on integrating DevilutionX into the official Debian repositories, so all Debian users could install it by running `apt install devilutionx` without having to go through GitHub. This would in turn make it available for Debian derivatives like Ubuntu or Linux Mint.

If you wish to follow all activity on this front, I will keep the following Debian bug report updated: https://bugs.debian.org/1088174

If some maintainers of DevilutionX would like to be kept in the loop for anything DevilutionX-related on the Debian side, please share an e-mail with me and I’ll make sure to include you in all further discussion.

Obviously, feel free to ask for any kind of detail, and I’ll answer to the best of my abilities.

**Comment by AJenbo** [comment]
Thanks, I tried having things like `game-data-packager` in mind when modeling something. It should be possible to fetch the shareware and extract spawn.mpq from it using smpq so that might be a useful option similar to how you can install the shareware data from doom and quake.

```
wget http://ftp.blizzard.com/pub/demos/diablosw.exe
smpq -x diablosw.exe spawn.mpq
```

If both spawn.mpq and diabdat.mpq are available the engine will prefer the full game.

**Reply by vv221** [reply]
To be fair this initial packaging is only going to include the engine, so Debian users are still going to be directed towards tools like game-data-packager or ./play.it to fetch and provide the game data. But now they will at least be able to fetch the engine directly from their distribution repositories, with the added benefit that these tools will no longer have to be updated to handle each new DevilutionX release, as they can expect the distribution to provide it already.

Please don’t bother too much about game-data-packager or ./play.it when designing DevilutionX, it’s our job on Debian side to ensure all things work well together. If really some future quirk makes it harder to work with DevilutionX, we’ll get in touch and see how to resolve that. But until then you can assume that all is in a good state.

---

Your suggestion about the shareware data is interesting, I wonder if we could provide that directly from the non-free section of the Debian repositories (legal questions might get in the way, so no promises on this one). So users would be one `apt install (…)` away to be able to play the shareware build of Diablo through DevilutionX, and would need to rely on extra operations only if they want to play the full game.

**Reply by AJenbo** [reply]
> Please don’t bother too much about game-data-packager or ./play.it when designing DevilutionX

Too late by a few years: https://github.com/diasurgical/devilutionX/issues/387#issuecomment-570090805

**Reply by AJenbo** [reply]
> I wonder if we could provide that directly from the non-free section of the Debian repositories

On the retail CD you get this pop up when selecting to install the shareware:

![image](https://github.com/user-attachments/assets/773843ec-9ffc-46f6-a9ca-b70fb13137bc)

> You can install a spawned version of Diablo on your friends' systems so that they can play Diablo without having the Diablo CD-ROM in their computer.

The original download page for the shareware: https://web.archive.org/web/19980128201335/http://www.blizzard.com/library/diablosw.htm

>  This demo is intended for private use only, and is not to be commercially distributed.

**Comment by AJenbo** [comment]
> While only the GOG installer is supported for now, adding support for the CD-ROM versions of the game
could be done too.

The CD is very simple, the diabdat.mpq is in the root, just copy it or point the game to it.

p.s. you mentioned not being in email contact with us, I have also fixed that.

**Reply by vv221** [reply]
I think I still have an original copy of the CD-ROM around (Diablo was actually one of my very first video games, a couple years after its release), I’m going to run some tests with that and see what I can get out of it.

**Reply by AJenbo** [reply]
Here is me running it directly from the original CD: https://www.youtube.com/watch?v=fk9GwVIhOGc

The 4 Hellfire MPQ files are also present directly on the Helfire CD

**Comment by AJenbo** [comment]
We have a guide for how to extract both the Diablo and Hellfire MPQ files from the GOG installer here in case that's helpful: https://github.com/diasurgical/devilutionX/wiki/Extracting-MPQs-from-the-GoG-installer

**Comment by AJenbo** [comment]
Any progress?

**Reply by vv221** [reply]
Not yet, I just finished working on another game engine (Julius, a free engine for Caesar 3).

The next one I’m going to work on should be either Jazz² Resurrection (free engine for Jazz Jackrabbit 2) or DevilutionX.

**Reply by krekhovx** [reply]
Hi, here are the problems I encountered when packaging:
https://github.com/diasurgical/DevilutionX/issues/7971

**Comment by vv221** [comment]
Sadly the first DevilutionX upload (based on version 1.5.4) got rejected because of license incompatibilities:
https://dfsg-new-queue.debian.org/reviews/devilutionx

For convenience, here is a copy of the final comment by the Debian reviewer, motivating the rejection:
> I took another close look at the package.
> 
> I found some more incompatibilities:
> 
> The Apache License, Version 1.0 is not compatible with all GNU GPL (in any version).
> The Apache License, Version 2.0 is not compatible with the GNU GPLv2 (only).
> The SustainableUse is a non-free (nc and nd) license which is not compatible with all GNU GPL (in any version).
> The MS-PL is also incompatible with the GNU GPL (in any version),
> 
> You must not insert code under GPL (in any version) in a non-free package

**Reply by StephenCWills** [reply]
Is this effort dead in the water?

I didn't think we had any GPL, but I guess ZeroTier pulls in some GPL dependencies. The one that surprised me, though, was libmpq. Hypothetically, you could package the game without ZT, but libmpq is not optional. I suppose libmpq would have to be submitted as a separate package to make it work.

**Reply by vv221** [reply]
> Is this effort dead in the water?

Absolutely not ;)

I’m not good with license stuff, but this is going to be the learning opportunity I was waiting for. From what I understand, the main incompatibility problem is due to vendored libraries, and could be avoided by packaging them separately and linking dynamically against them during build time?

This is something I can work on, starting with libmpq.

If some libraries would be harder to package and DevilutionX can be built without them (at the cost of some features), let us do that to get it in. We can then go back to these libraries and take the needed time to get them packaged, and this way restore all of DevilutionX features in Debian.

And if by anything I wrote here I give the impression of misunderstanding something, it probably means I do indeed misunderstand. Feel free to correct me, I know I’m way outside of my usual skills set here.

---

Before taking any real initiative I’m going to let some time for @krekhovx to react, and see how we can share the work to ensure we don’t duplicate some tasks.

---

TL;DR: I *want* DevilutionX in Debian, and am going to put all required effort to ensure it happens.

**Reply by glebm** [reply]
> could be avoided by packaging them separately and linking dynamically against them during build time?

The only mandatory vendored libraries are the ones that DevilutionX had to fork, they're not equivalent to and so can't be swapped in to upstream versions of these libraries. While we tried to upstream our patches (e.g. to libzt and libmpq), it wasn't always possible (inactive projects etc)

**Reply by krekhovx** [reply]
> Before taking any real initiative I’m going to let some time for @krekhovx to react, and see how we can share the work to ensure we don’t duplicate some tasks.

I spent quite a lot of time packaging DevilutionX, so I’m admittedly a bit disappointed that it didn’t get accepted yet. At the moment I’m busy with other Debian packages, but I’m absolutely willing to help where needed to move this forward. Unfortunately, I won’t be able to dedicate much time to it right now, so I would really appreciate it if you could take the lead on most of the remaining work.

**Reply by smcv** [reply]
I'm a Debian contributor but not a member of the DFSG team, and was not involved in this decision. I would like to find a way to have this engine in Debian, but I'm not sure that I can see how, under current licensing.

> The SustainableUse is a non-free (nc and nd) license which is not compatible with all GNU GPL (in any version).

If DevilutionX contains both SUL-1.0 and GPL components, then I think that's going to be the key licensing problem. Briefly, the GPL can be summarized as "must make the whole source of derivative works available under the GPL", but SUL-1.0 is not GPL-compatible (it has the additional restriction "for non-commercial purposes" which is a "further restriction" in GPL-2.0 §6 or GPL-3.0 §10), so you can't comply with both licenses at the same time. This is not really Debian's choice: it isn't about Debian's self-imposed policies, but rather about complying with license terms to avoid copyright infringement. A redistributor like Debian has to obey both what the libmpq authors require and what the DevilutionX authors require, and if we can't do both at the same time, then we aren't allowed to redistribute at all.

If (some of?) the GPL code is (a fork of) libmpq, unfortunately packaging it separately is not going to solve this, because a copy of DevilutionX that is dynamically linked to an external libmpq is still believed to be a derivative work of libmpq, and therefore distributing it under a non-GPL-compatible license would be seen as infringing the libmpq authors' copyright, and not legal for Debian or other distros to ship.

(I'm saying "believed to be" because what is or isn't a derivative work is a question of copyright law rather than what any individual license says, and as far as I know there has never been a court ruling on whether dynamic linking creates a derivative work, but the FSF claims that it does and it seems that industry consensus is that we believe them. Also, if DevolutionX is sufficiently closely-integrated with libmpq that it needs a specific modified version, I think it would be difficult to claim that some of its functionality *isn't* derived from libmpq.)

Similarly the MS-PL is listed on https://www.gnu.org/licenses/license-list.html as incompatible with the GPL (I haven't looked into specifically why, but presumably one of its terms is considered to be a "further restriction") so that combination would also be considered copyright-infringing; and similarly the Apache licenses have terms that are treated as "further restrictions" by the GPL. Sorry, but some licenses just aren't compatible, and can't be combined in one work without special permission from the copyright holders.

Are all of these licenses "set in stone", or is there a possibility of some of the components being relicensed by their copyright holders?

**Reply by StephenCWills** [reply]
It seems I incorrectly assumed this was an issue regarding redistrIbution of GPL source code in a package with an incompatible license. I hadn't considered the issue also extends to linking. That is indeed a problematic restriction.

Unfortunately, changing the license used by DevilutionX is not an option. Although the libmpq project hasn't been touched in 15 years, it does look like the author of libmpq was active on GitHub in the last few months so it may be possible to get in touch with him.

**Reply by AJenbo** [reply]
>  Also, if DevolutionX is sufficiently closely-integrated with libmpq that it needs a specific modified version, I think it would be difficult to claim that some of its functionality isn't derived from libmpq.)

We wanted to migrate from StormLib to libmpq, but libmpq was missing some features which we implemented, so I would say it's more the case that libmpq is the derivative, and since DevilutionX originally used StormLib I would say it's not modeled after it. The problem with StormLib is that it isn't portable code, if you want you could make DevilutionX able to link with either and use StormLib on platforms where it is available and works correctly. StormLib upstream does have all required features so we don't require a modified version of it.

Since some of libmpq's code is adapted from StormLib should it present those files under the MIT?

> Are all of these licenses "set in stone"

Our license requirement is:
- As open as possible
- Has a "for non-commercial purposes" clause

This is per agreement with some people with real lawyers.

I don't know if libmpq is willing to change theirs, last update was 15 years and they haven't responded to any of our contributions which is why we can't use the upstream version.

**Reply by smcv** [reply]
> Our license requirement is: [...] Has a "for non-commercial purposes" clause

I can take a reasonably confident guess at why this is a requirement; but, sorry, this restriction isn't GPL-compatible, so GPL'd library code isn't a suitable dependency (unless the copyright holders of the GPL'd code are willing to relicense, but then your copy of the library isn't GPL'd code any more).

LGPL code that is dynamically linked and therefore could be swapped out for a modified version (DLL or shared library) would be OK, if I understand correctly - this is the main difference between "strong copyleft" like the GPL, and "weak copyleft" like the LGPL.

> The problem with StormLib is that it isn't portable code, if you want you could make DevilutionX able to link with either and use StormLib on platforms where it is available and works correctly

Debian is only interested in Linux (and maybe Hurd as an unofficial port but not many people use or maintain that), and Stormlib seems to support (only) Windows and Linux if I'm reading correctly, so that wouldn't be a problem from Debian's point of view - indeed, [it's already in Debian](https://tracker.debian.org/pkg/stormlib). @krekhovx, would you be interested in looking into teaching DevilutionX to be able to use either library, at some point?

Are there other GPL dependencies? If https://salsa.debian.org/games-team/devilutionx/-/blob/debian/latest/debian/copyright?ref_type=heads is accurate, it mentions files under `dist/libzt-src/ext/ZeroTierOne/attic/historic/anode/` and `dist/libzt-src/ext/ZeroTierOne/java/jni/`, `dist/libzt-src/ext/ZeroTierOne/ext/bin/tap-windows-ndis6/`, `dist/libzt-src/ext/ZeroTierOne/windows/TapDriver6/`, `dist/libzt-src/ext/ZeroTierOne/ext/libnatpmp/fr/free/miniupnp/libnatpmp/`, `dist/libzt-src/ext/ZeroTierOne/java/`, etc. but perhaps those aren't actually necessary? (Typically if there are files with awkward licensing that aren't actually necessary, a Debian package would delete them before doing the build, as a convenient way to prove that they aren't being used.)

Similarly if `dist/libzt-src/ext/ZeroTierOne/windows/ZeroTierOne/` isn't relevant on Linux, the Debian package build could just delete it, and from Debian's point of view that would make questions about MS-PL license compatibility go away.

> Since some of libmpq's code is adapted from StormLib should it present those files under the MIT?

Removing the MIT licensing information from MIT-licensed code isn't allowed ("shall be included in all copies or substantial portions of the Software") but because the MIT license isn't a "copyleft" license like the GPL, it's valid to have something that is a combination of MIT code plus new code, and put the result under the GPL. It's quite common for (L)GPL projects to incorporate something under a permissive license like this. Typically this is represented by having the (L)GPL license grant at the top of the file, and then saying something like "parts of this file based on Stormlib, under this license: ..." and quoting Stormlib's copyright holders and MIT license grant.

**Reply by smcv** [reply]
> It's quite common for (L)GPL projects to incorporate something under a permissive license like this

The first example I found is https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gutils.c?ref_type=heads, which is LGPL rather than GPL but the principle is the same: if you look for the definition of `load_user_special_dirs_unlocked`, you'll see that it's MIT-licensed code taken from somewhere else, and whoever integrated it into GLib has acknowledged that.

**Reply by StephenCWills** [reply]
> Are there other GPL dependencies?

Those other files are all dependencies of libzt, which is necessary for playing Multiplayer on the ZeroTier network. Many platforms that we support aren't supported by libzt so it is already an optional component of DevilutionX at build time. At least initially, I assume it would be wise to focus on resolving the issue with libmpq since we can simply disable ZeroTier for now and deal with it separately later.

**Reply by AJenbo** [reply]
> Debian is only interested in Linux (and maybe Hurd

Afaik StormLib has issues on ARM, it's compression is hosed when build for non-x86.

**Reply by AJenbo** [reply]
> Removing the MIT licensing information from MIT-licensed code isn't allowed

Right, that was my point. I'm guessing the author of libmpq didn't intend any of these issues.

@mbroemme if you see this would you be willing to consider re-licensing libmpq under LGPL instead of GPL so that it can be used as a lib by projects that aren't themselves GPL, or is this a matter of principle to you that only other GPL project may use it even when we contribute improvements back to libmpq?

**Reply by smcv** [reply]
> Afaik StormLib has issues on ARM

Oh, I see, the other axis of portability: between CPUs/architecture ABIs, rather than between OS ABIs.

**Reply by glebm** [reply]
Since mbroemme@'s libmpq hasn't been updated for **15 years**, another option would be to say our version of libmpq is the new mainline. Then, after the Debian package for libmpq is updated to be based on our version, we can simply link it dynamically for the Debian distro.

**Reply by smcv** [reply]
> Then, after the Debian package for libmpq is updated to be based on our version, we can simply link it dynamically for the Debian distro

That doesn't solve the license incompatibility, though: the conventional interpretation of the GPL is that the dynamically-linked DevilutionX binary would be a derivative work of both DevilutionX source code (which is SUL-1.0) and libmpq source code (which is under some version of the GPL), which means the GPL says "you must release it under GPL", but the SUL-1.0 says "you can't release it under GPL", and the result is that it can't be released at all.

**Reply by smcv** [reply]
If the copyright holders of libmpq gave permission to relicense it under some suitable version of the LGPL as suggested in https://github.com/diasurgical/DevilutionX/discussions/7556#discussioncomment-15924708, *then* it would be OK, as long as it's dynamically-linked so that it can make use of LGPL-2.1 paragraph 6.b or LGPL-3.0 paragraph 4.d.1. But without that relicensing, it's the GPL's terms that apply, and those are sometimes inconveniently strict.

As far as I can see it doesn't make any legal difference whether the version of libmpq in use is bundled with DevilutionX or packaged separately.

**Reply by AJenbo** [reply]
@glebm @StephenCWills can you please evaluate this solution (burn it with fire is a valid option):

https://github.com/diasurgical/DevilutionX/pull/8482

**Reply by mbroemme** [reply]
> > Removing the MIT licensing information from MIT-licensed code isn't allowed
> 
> Right, that was my point. I'm guessing the author of libmpq didn't intend any of these issues.
> 
> @mbroemme if you see this would you be willing to consider re-licensing libmpq under LGPL instead of GPL so that it can be used as a lib by projects that aren't them selfs GPL, or is this a matter of principal to you that only other GPL project may use it even when we contribute improvements back to libmpq?

It's actually not a problem to re-license it under the LGPL as I'm still the copyright holder. I know I haven't updated the library in a long time, but this might be a good moment to pick it up again. 🙂 Back then I was a bit too strict about linking it in commercial products, but my view has changed over time.

**Reply by AJenbo** [reply]
We have now migrated to mpqfs which is a MIT licensed MPQ 1.0 reader / writer purpose build for our needs. That should both improve performance and avoid the license issue.

**Reply by mbroemme** [reply]
> > Removing the MIT licensing information from MIT-licensed code isn't allowed
> 
> Right, that was my point. I'm guessing the author of libmpq didn't intend any of these issues.
> 
> @mbroemme if you see this would you be willing to consider re-licensing libmpq under LGPL instead of GPL so that it can be used as a lib by projects that aren't themselves GPL, or is this a matter of principle to you that only other GPL project may use it even when we contribute improvements back to libmpq?

It is done and has been re-licensed under LGPL with permission of all contributors, more information at https://github.com/mbroemme/libmpq/blob/master/RELICENSING.md and v0.5.0 has been released https://github.com/mbroemme/libmpq/releases/tag/v0.5.0

@glebm I will also review your pending PRs in the next few days and more features for newer MPQ formats will be added as well.

**Comment by vv221** [comment]
> I spent quite a lot of time packaging DevilutionX, so I’m admittedly a bit disappointed that it didn’t get accepted yet.

That’s understandable. I was nothing short of furious when I realised the current Debian DFSG Team was actually inventing new reasons on the fly to reject DevilutionX once I proposed resolutions. They went as far as pretending that only *free/libre* software is allowed in Debian *non-free*!

For now I no longer want to interact with them, they are clearly not qualified for their role and I am not going to waste more time with them. Actually it piled on top of other things and led to me leaving Debian (both as a contributor and as an user), so do not expect further activity from me about integrating new software in Debian.

Gentoo has had DevilutionX packaged for years, so this is the distribution I am going to use on my main computers from now on.

**Reply by krekhovx** [reply]
Wow, Antoine, I’m really sorry to hear that you’ve decided to leave Debian.

You’ve made a significant contribution over the years, and I truly appreciate the work you’ve done. I was honestly surprised when you stepped away from your packages, but now I understand the context much better. Yes, this really does seem like an unfair situation.

> so do not expect further activity from me about integrating new software in Debian.

Okey :( We will miss you!
